Event unsubscription from DeviceProxy destructor

mliszcz opened the pull request:
The pointer to device proxy used for event subscription is now stored in
EventSubscribeStruct (which is unique entry per subscription) instead of
EventCallBackStruct (which is shared between all subscriptions to the
same event).

Ability to associate subscriptions with device proxies is needed to
correctly unsubscribe from events inside device proxy destructor.

Fixes #353

mliszcz committed:
Add test with two proxies to the same event

Tests that when two or more device proxies subscribe to the same event
and one of them is deleted, only subscriptions made with that proxy are
cancelled.

mliszcz committed:
Track event subscriptions by device proxy pointer

The pointer to device proxy used for event subscription is now stored in
EventSubscribeStruct (which is unique entry per subscription) instead of
EventCallBackStruct (which is shared between all subscriptions to the
same event).

Ability to associate subscriptions with device proxies is needed to
correctly unsubscribe from events inside device proxy destructor.

t-b reviewed (dismissed):
LGTM. It took me a while to understand that `callback_list[0]` is always valid. But in https://github.com/tango-controls/cppTango/blob/d067b589985c701ae49541f3a27366ceaab0f196/cppapi/client/event.cpp#L1745 we always add one entry.

mliszcz commented:
Thanks @t-b for this comment. I agree it is a bit fishy. Maybe I'll add assertion to all the places to clearly indicate that the presence of at least one entry is guaranteed? What do you think?
```cpp
assert(!epos->second.callback_list.empty());
// access callback_list[0]
```

t-b commented:
@mliszcz An assertion yes, but maybe only once? Can you add a wrapper function which returns `callback_list[0]` and then only that wrapper has the assertion? But I'm also okay with as it is.

mliszcz committed:
Ensure that EventCallBackStruct::callback_list is never empty

t-b commented:
Automatic unsubscription in DeviceProxy dtor was added in 36281159 (- Add automatic event un-subscription in DeviceProxy class dtor, 2016-06-10).

mliszcz commented:
HI @t-b can you re-check the last commit (the one changing access to `callback_list[0]`)? It dismissed your previous review.

t-b reviewed (dismissed):
LGTM

bourtemb reviewed (commented):
Thanks @mliszcz for the work on this PR.

It looks like there is one case which is not well handled here.
When a user subscribes to the same event (same attribute and same event type) several times using the same DeviceProxy but using different callbacks.

If the DeviceProxy object is deleted, it should unsubscribe all the subscriptions done with this DeviceProxy object.
With the current solution in this PR, during DeviceProxy deletion, it unsubscribes only for the 1st callback in the use case I described it seems and some callbacks are still called when they should not after DeviceProxy deletion.

mliszcz committed:
Store proxy in new subscriptions to already subscribed events

EventSubscribeStruct::device must always be initialized to correctly
unsubscribe from all the events when the device proxy is deleted.

mliszcz commented:
Thanks @bourtemb for catching this.There was indeed one place where I forgot to initialize the new DeviceProxy* field in event subscription struct. It is fixed now and test is added in the latest commit.

bourtemb reviewed (approved):
Thanks @mliszcz !
It seems to work as well for the use case I was describing now.
Thanks for the new test for that too!
Looks good to me!
